### PR TITLE
Sort fields and options before generating reference

### PR DESCRIPTION
**Pull request by tgolsson**
Noticed in #139 that fields and options were jumping around with no reason. After some testing, I've concluded that the JSON output from `pants help-all` is non-deterministically ordered. Since the JS just loads that data, it ends up shifting around the docs as well.

I had a quick peek in the help-all code and while I think it would be possible to affect it there, this seems easier/quicker to fix (and only pays the price here). The downside is that this is a bigger diff than the spurious changes that now occur.

I did a second commit where I  played around with also sorting required fields before optional fields, which is what the CLI seems to do. If we do that, we should probably also annotate required fields in some way. I think there's pros and cons with both.

Let me know what you think.

### CODE REVIEW

File: reference_codegen/generate.mjs [hunk 1]
```diff
@@ -258,13 +258,28 @@ Object.entries(helpAll.scope_to_help_info).forEach(([scope, info]) => {
...
 Object.entries(helpAll.name_to_target_type_info).forEach(([name, info]) => {
   info.fields.forEach((field) => {
     field.default = convertDefault(field.default, field.typ);
     field.description = convertDescription(field.description);
   });
+
+  // sort the fields to make the output deterministic
+  info.fields.sort((a, b) => {
+    if (a.required == b.required) {
+      return a.alias.localeCompare(b.alias);
```
Comment (thejcannon) [inline reference_codegen/generate.mjs:276]: Is there a reason you dont just return unconditionally after the two `if` statements below?

**Comment by tgolsson** [comment]
(I skipped regen docs because PR would be humongous will do if wanted or obviously before merge.)

**Review by thejcannon — APPROVED** [review]
This is an instance I wish I had PR previews.

I assume you tested this and it looks alright?

> If we do that, we should probably also annotate required fields in some way. I think there's pros and cons with both.

Can you elaborate? I thought required fields _did_ say they were required.

**Comment by tgolsson** [comment]
> Can you elaborate? I thought required fields did say they were required.

Oh, they sort of do I guess. We say required if they don't have a default_repr, which I guess is the same thing. We don't actually look at the required property itself.

File: reference_codegen/generate.mjs (hunk 1 above)
Comment (thejcannon) [reply reference_codegen/generate.mjs:276]: Just wanted to make sure I didn't miss anything 😅

**Comment by tgolsson** [comment]
@thejcannon PTAL again, added a fix for the default value. You can see the changed sort + fixed default value below, which seems OK to me. Not sure if the sorting change is worth getting some opinions on on Slack?

image

**Review by thejcannon — APPROVED** [review]

**Pull request merged**